ethers-contract/src/stream.rs

```
@@ -49,9 +53,140 @@ where
+    ///
+    /// #[derive(Clone, Debug, EthEvent)]
+    /// pub struct Approval {
+    ///     #[ethevent(indexed)]
+    ///     pub token_owner: Address,
+    ///     #[ethevent(indexed)]
+    ///     pub spender: Address,
+    ///     pub tokens: U256,
+    /// }
+    ///
+    /// #[derive(Clone, Debug, EthEvent)]
+    /// pub struct Transfer {
+    ///     #[ethevent(indexed)]
+    ///     pub from: Address,
+    ///     #[ethevent(indexed)]
+    ///     pub to: Address,
+    ///     pub tokens: U256,
+    /// }
+    ///
+    ///
+    /// let ev1 = contract.event::<Approval>().from_block(1337).to_block(2000);
+    /// let ev2 = contract.event::<Transfer>();
+    ///
+    /// let mut events = ev1.stream().await.unwrap().select(ev2.stream().await.unwrap()).ok();
+    ///
+    /// while let Some(either) = events.next().await {
+    ///     match either {
+    ///         Either::Left(approval) => { let Approval{token_owner,spender,tokens} = approval; }
+    ///         Either::Right(transfer) => { let Transfer{from,to,tokens} = transfer; }
+    ///     }
+    /// }
+    ///
+    /// # }
+    /// ```
+    pub fn select<St>(self, st: St) -> SelectEvent<SelectEither<'a, Result<R, E>, St::Item>>
+    where
+        St: Stream + Unpin + 'a,
+    {
+        SelectEvent(Box::pin(futures_util::stream::select(
+            self.map(Either::Left),
+            st.map(Either::Right),
+        )))
+    }
+}
+
+pub type SelectEither<'a, L, R> = Pin<Box<dyn Stream<Item = Either<L, R>> + 'a>>;
+
+#[pin_project]
+pub struct SelectEvent<T>(#[pin] T);
+
+impl<'a, T, L, LE, R, RE> SelectEvent<T>
+where
+    T: Stream<Item = Either<Result<L, LE>, Result<R, RE>>> + 'a,
+    L: 'a,
+    LE: 'a,
+    R: 'a,
+    RE: 'a,
+{
+    /// Turns a stream of Results to a stream of `Result::ok` for both arms
+    pub fn ok(self) -> Pin<Box<dyn Stream<Item = Either<L, R>> + 'a>> {
+        Box::pin(self.filter_map(|e| async move {
```

Do we want the `filter_map` here? Won't this silently eat all errors if any are observed?

---

yes this skips all erroneous values, but is probably useful if you're only interested in the `ok` values.